Sources/Private/Experimental/Layers/GradientRenderLayer.swift

```diff
@@ -0,0 +1,86 @@
+// Created by Cal Stephens on 1/10/22.
+// Copyright © 2022 Airbnb Inc. All rights reserved.
+
+import UIKit
+
+// MARK: - GradientRenderLayer
+
+/// A `CAGradientLayer` subclass used to render a gradient _outside_ the normal layer bounds
+///
+///  - `GradientFill.startPoint` and `GradientFill.endPoint` are expressed
+///    with respect to the `bounds` of the `ShapeItemLayer`.
+///
+///  - The gradient itself is supposed to be rendered infinitely in all directions
+///    (e.g. including outside of `bounds`). This is because `ShapeItemLayer` paths
+///    don't necessarily sit within the layer's `bounds`.
+///
+///  - To support this, `GradientRenderLayer` tracks a `gradientRenderBounds`
+///    that `startPoint` / `endPoint` are calculated relative to.
+///    The _actual_ `bounds` of this layer is padded by a large amount so that
+///    the gradient can be drawn outside of the `gradientRenderBounds`.
+///
+final class GradientRenderLayer: CAGradientLayer {
```

Comment:
This subclass lets us properly support drawing gradients outside the `bounds` of the `ShapeItemLayer`, which is necessary because the `path` of the `CAShapeLayer` isn't necessarily within the `bounds` of its parent layer.

In the left image, you can see what happens if you just use `CAGradientLayer` (the top-left corner of the small square is `bounds.origin` (`CGPoint.zero`). This subclass lets us draw the gradient above and to the left of that bounds rectangle, as if the gradient itself is "infinite" (that's the behavior of `CGContext.drawLinearGradient` with `[.drawsAfterEndLocation, .drawsBeforeStartLocation]`, which is used by the existing rendering engine).

| Without this subclass | With this subclass |
| ----- | ---- |
| ![reference_1_1658A012-E8C2-4AC3-87E9-B6A482880F3F](https://user-images.githubusercontent.com/1811727/148859341-9894344e-026e-4264-8343-eadb5c43bff2.png) | ![failure_2_1658A012-E8C2-4AC3-87E9-B6A482880F3F](https://user-images.githubusercontent.com/1811727/148859395-69f22378-1ee5-46af-933b-409316a14b22.png) |